deno/private/binary.bzl

```diff
@@ -17,32 +17,57 @@ def _deno_binary_impl(ctx):
     outfile = ctx.actions.declare_file(outfile_name)
 
     # Get Deno executable and build flags to include in launcher
-    deno = ctx.toolchains["@aspect_rules_deno//deno:toolchain_type"]
-    flags = _build_deno_flags(unstable_apis = ctx.attr.unstable_apis, allow = ctx.attr.allow)
+    deno_tools = ctx.toolchains["@aspect_rules_deno//deno:toolchain_type"].denoinfo.tool_files
+    flags = _build_deno_flags(
+        allow_uncached = ctx.attr.allow_uncached,
+        unstable_apis = ctx.attr.unstable_apis,
+        allow = ctx.attr.allow,
+    )
     runtime_args = leftover_args
 
-    # Set DENO_DIR within build bin
-    deno_executable = deno.denoinfo.tool_files[0].short_path
-    deno_cache = deno_executable + "_cache"
+    # Add lock file to flags if specified
+    build_flags = []
+    if ctx.file.lockfile != None:
+        build_flags.append("--lock")
+        build_flags.append(ctx.file.lockfile.path)
+    flags += build_flags
+
+    # Create cache folder for use as DENO_DIR
+    cache_folder = ctx.actions.declare_directory("%s_deno_cache" % ctx.attr.name)
+
+    # Collect up all of the input files for use in actions
+    infiles = ctx.files.main + ctx.files.deps + ctx.files.lockfile
+
+    # Download any remote included scripts to cache
+    ctx.actions.run(
+        executable = deno_tools[0].path,
+        arguments = ["cache", ctx.file.main.path] + build_flags,
+        tools = deno_tools,
+        env = {
+            "DENO_DIR": cache_folder.path,
+        },
+        mnemonic = "DenoCache",
+        inputs = infiles,
+        outputs = [cache_folder],
```

Comment:
probably want to add the `requires-network` execution requirement here... although remote execution environment that are network sandboxed don't seem to honour it it seems